src/GdbServer.cc

```diff
@@ -1806,6 +1808,56 @@ static bool is_likely_interp(string fsname) {
   return fsname == "/lib64/ld-linux-x86-64.so.2" || fsname == "/lib/ld-linux.so.2";
 }
 
+static remote_ptr<void> base_addr_from_rendezvous(Task* t, string fname)
+{
+  std::vector<uint8_t> auxv = t->vm()->saved_auxv();
+  if (auxv.size() == 0 || (auxv.size() % sizeof(uint64_t) != 0)) {
+    // Corrupted or missing auxv
+    return nullptr;
+  }
+  remote_ptr<void> interpreter_base = nullptr;
+  bool found = false;
+  for (size_t i = 0; i < (auxv.size() / sizeof(uint64_t)) - 1; i += 2) {
+    uint64_t* entry = ((uint64_t*)auxv.data())+i;
+    uint64_t kind = entry[0];
+    uint64_t value = entry[1];
+    if (kind == AT_BASE) {
+      interpreter_base = value;
+      found = true;
+      break;
+    }
+  }
+  if (!found) {
+    return nullptr;
+  }
+  string ld_path = t->vm()->mapping_of(interpreter_base).map.fsname();
+  ScopedFd ld(ld_path.c_str(), O_RDONLY);
+  ElfFileReader reader(ld);
+  auto syms = reader.read_symbols(".dynsym", ".dynstr");
+  static const char r_debug[] = "_r_debug";
+  found = false;
+  uintptr_t r_debug_offset = 0;
+  for (size_t i = 0; i < syms.size(); ++i) {
+    if (!syms.is_name(i, r_debug)) {
+      continue;
+    }
+    r_debug_offset = syms.addr(i);
+    found = true;
+  }
+  if (!found) {
+    return nullptr;
+  }
+  remote_ptr<NativeArch::r_debug> r_debug_remote = interpreter_base.as_int()+r_debug_offset;
+  remote_ptr<NativeArch::link_map> link_map = t->read_mem(REMOTE_PTR_FIELD(r_debug_remote, r_map));
+  while (link_map != nullptr) {
+    if (fname == t->read_c_str(t->read_mem(REMOTE_PTR_FIELD(link_map, l_name)))) {
+      return t->read_mem(REMOTE_PTR_FIELD(link_map, l_addr));
+    }
+    link_map = t->read_mem(REMOTE_PTR_FIELD(link_map, l_next));
```

Comment:
I think the approach is fine but I think these `read_mem`s need to handle failure gracefully in case the link map is corrupted.

Reply:
done